### DISCUSSION TITLE
When does the high level consumer actually call poll?

### DISCUSSION THREAD

**Discussion by JSoet** [edenhill/librdkafka#4095] [Q&A] [answered]
We are using the high level consumer and so we use the `consume` method to get messages, and I was wondering when it actually does the polling...

I assume that calling `consume` will trigger a poll to the broker if there are currently no messages stored in memory. But will poll be called at any other time (in other words, will librdkafka automatically poll in the background to ensure that there is a "buffer" of messages? I thought I remembered reading this at some point, but can't find it now, maybe it was just an assumption on my end)... Or does it just `poll` the broker only when `consume` is called?

**Comment by edenhill** [comment, marked as answer]
This means that consume() and consumer_poll() (et.al) consume() does not have any side effects, it simply pops messages off the internal prefetch queue.


Messages are prefetched in backgrounds threads to achieve the queued.min.messages (et.al) configuration properties.
As the queue count drops below the configured thresholds, more messages will be pre-fetched from the brokers.

https://github.com/edenhill/librdkafka/wiki/FAQ#how-are-partitions-fetched

**Reply by JSoet** [reply]
Thanks a lot for the prompt reply, was working on librdkafka more a while back, and didn't find that now when I was looking for it.
